### ISSUE TITLE
Licenses fail to be loaded (NPE) when XWiki is restarted

### ISSUE THREAD

**Issue by Enygma2002** [xwikisas/application-licensing#52]
On XWiki 9.2, when getting a trial license for a paid app (for example) and then restarting XWiki, I get the following stacktrace and the licenses UI tells me I have no license:
```
2018-11-01 16:51:28,401 [main] WARN  c.x.l.i.DefaultLicenseManager  - Error retrieving license, license has been skipped. 
org.xwiki.component.manager.ComponentLookupException: Failed to lookup component [org.xwiki.instance.internal.DefaultInstanceIdManager] identified by [role = [interface org.xwiki.instance.InstanceIdManager] hint = [default]]
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:394)
        at org.xwiki.component.embed.EmbeddableComponentManager.getInstance(EmbeddableComponentManager.java:190)
        at org.xwiki.component.embed.GenericProvider.getInstance(GenericProvider.java:111)
        at org.xwiki.component.embed.GenericProvider.get(GenericProvider.java:100)
        at com.xwiki.licensing.internal.DefaultLicenseValidator.isApplicable(DefaultLicenseValidator.java:83)
        at com.xwiki.licensing.internal.DefaultLicenseManager.linkLicenseToLicensedFeature(DefaultLicenseManager.java:192)
        at com.xwiki.licensing.internal.DefaultLicenseManager.initialize(DefaultLicenseManager.java:122)
        at org.xwiki.component.embed.InitializableLifecycleHandler.handle(InitializableLifecycleHandler.java:39)
        at org.xwiki.component.embed.EmbeddableComponentManager.createInstance(EmbeddableComponentManager.java:322)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:424)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:392)
        at org.xwiki.component.embed.EmbeddableComponentManager.getInstance(EmbeddableComponentManager.java:190)
        at org.xwiki.component.embed.EmbeddableComponentManager.getDependencyInstance(EmbeddableComponentManager.java:363)
        at org.xwiki.component.embed.EmbeddableComponentManager.createInstance(EmbeddableComponentManager.java:312)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:424)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:392)
        at org.xwiki.component.embed.EmbeddableComponentManager.getInstance(EmbeddableComponentManager.java:190)
        at org.xwiki.component.embed.EmbeddableComponentManager.getDependencyInstance(EmbeddableComponentManager.java:363)
        at org.xwiki.component.embed.EmbeddableComponentManager.createInstance(EmbeddableComponentManager.java:312)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:424)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:392)
        at org.xwiki.component.embed.EmbeddableComponentManager.getInstance(EmbeddableComponentManager.java:190)
        at org.xwiki.observation.internal.DefaultObservationManager.onEventListenerComponentAdded(DefaultObservationManager.java:359)
        at org.xwiki.observation.internal.DefaultObservationManager.onComponentEvent(DefaultObservationManager.java:337)
        at org.xwiki.observation.internal.DefaultObservationManager.notify(DefaultObservationManager.java:283)
        at org.xwiki.component.internal.StackingComponentEventManager.sendEvent(StackingComponentEventManager.java:142)
        at org.xwiki.component.internal.StackingComponentEventManager.flushEvents(StackingComponentEventManager.java:92)
        at org.xwiki.extension.jar.internal.handler.JarExtensionHandler.loadComponents(JarExtensionHandler.java:234)
        at org.xwiki.extension.jar.internal.handler.JarExtensionHandler.install(JarExtensionHandler.java:164)
        at org.xwiki.extension.jar.internal.handler.JarExtensionHandler.initialize(JarExtensionHandler.java:142)
        at org.xwiki.extension.handler.internal.DefaultExtensionHandlerManager.initialize(DefaultExtensionHandlerManager.java:154)
        at org.xwiki.extension.handler.internal.DefaultExtensionInitializer.initializeExtensionInNamespace(DefaultExtensionInitializer.java:194)
        at org.xwiki.extension.handler.internal.DefaultExtensionInitializer.initializeExtensionInNamespace(DefaultExtensionInitializer.java:159)
        at org.xwiki.extension.handler.internal.DefaultExtensionInitializer.initializeExtension(DefaultExtensionInitializer.java:131)
        at org.xwiki.extension.handler.internal.DefaultExtensionInitializer.initialize(DefaultExtensionInitializer.java:103)
        at org.xwiki.extension.handler.internal.DefaultExtensionInitializer.initialize(DefaultExtensionInitializer.java:79)
        at org.xwiki.extension.handler.internal.ExtensionApplicationStartedListener.onEvent(ExtensionApplicationStartedListener.java:76)
        at org.xwiki.observation.internal.DefaultObservationManager.notify(DefaultObservationManager.java:304)
        at org.xwiki.observation.internal.DefaultObservationManager.notify(DefaultObservationManager.java:269)
        at org.xwiki.observation.internal.DefaultObservationManager.notify(DefaultObservationManager.java:321)
        at org.xwiki.container.servlet.XWikiServletContextListener.contextInitialized(XWikiServletContextListener.java:118)
        at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:800)
        at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:444)
        at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:791)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:294)
        at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1349)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1342)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:741)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:505)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.deploy.bindings.StandardStarter.processBinding(StandardStarter.java:41)
        at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:186)
        at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:498)
        at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:146)
        at org.eclipse.jetty.deploy.providers.ScanningAppProvider.fileAdded(ScanningAppProvider.java:180)
        at org.eclipse.jetty.deploy.providers.WebAppProvider.fileAdded(WebAppProvider.java:461)
        at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.fileAdded(ScanningAppProvider.java:64)
        at org.eclipse.jetty.util.Scanner.reportAddition(Scanner.java:609)
        at org.eclipse.jetty.util.Scanner.reportDifferences(Scanner.java:528)
        at org.eclipse.jetty.util.Scanner.scan(Scanner.java:391)
        at org.eclipse.jetty.util.Scanner.doStart(Scanner.java:313)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.deploy.providers.ScanningAppProvider.doStart(ScanningAppProvider.java:150)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.deploy.DeploymentManager.startAppProvider(DeploymentManager.java:560)
        at org.eclipse.jetty.deploy.DeploymentManager.doStart(DeploymentManager.java:235)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:132)
        at org.eclipse.jetty.server.Server.start(Server.java:387)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:114)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
        at org.eclipse.jetty.server.Server.doStart(Server.java:354)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.xml.XmlConfiguration$1.run(XmlConfiguration.java:1255)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1174)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.jetty.start.Main.invokeMain(Main.java:321)
        at org.eclipse.jetty.start.Main.start(Main.java:817)
        at org.eclipse.jetty.start.Main.main(Main.java:112)
Caused by: java.lang.NullPointerException: null
        at org.xwiki.instance.internal.DefaultInstanceIdManager.getXWikiContext(DefaultInstanceIdManager.java:151)
        at org.xwiki.instance.internal.DefaultInstanceIdManager.initialize(DefaultInstanceIdManager.java:93)
        at org.xwiki.component.embed.InitializableLifecycleHandler.handle(InitializableLifecycleHandler.java:39)
        at org.xwiki.component.embed.EmbeddableComponentManager.createInstance(EmbeddableComponentManager.java:322)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:424)
        at org.xwiki.component.embed.EmbeddableComponentManager.getComponentInstance(EmbeddableComponentManager.java:392)
        ... 82 common frames omitted
```

The Licenses UI does, however, display my instance ID, so the InstanceIDManager eventually succeeds to initialize, just not at the time when the LicenseManager is initialized.

**Comment by evalica** [comment]
I've also reproduced this issue when installing Forum and Ideas apps with EM, on XWiki 11.0.3. The Forum app was fully functioning without the license. Ideas was accessible, but the livetable was not loading. 
The Licensor and messages about invalid licenses went back to normal when the server restarted. 

Some steps that I've done that might be of interest: 
* the installation was done with an Admin user, but not the default one;
* I've segmented the installation process, meaning after pressing on "Install", I've changed views, going to Home, etc.; coming back to press on "Continue"; and I wasn't in EM when the installation finished. When the install is complete, I've accessed the app though the AppBar directly. 
* the Forum's app translation key for the AppBar entry was broken / not registered. It came back when accessing Forums space.

**Comment by mflorea** [comment]
Fixed by f4357a0b6d650fb457d97c6a56bfe153d87e2e1a for #57 . We now initialize the licensing system after the database is ready (when the ``ApplicationReadyEvent`` is fired) in order to make sure the ``InstanceIdManager`` can be used (the instance id needs to be read from the database).
